faunadb-java-dsl/src/main/java/com/faunadb/client/types/Decoder.java

```diff
@@ -0,0 +1,309 @@
+
+import java.lang.reflect.*;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB {@link Value} to object decoder.
+ */
+public class Decoder {
+    private final static ConcurrentHashMap<Class<?>, Creator> CREATORS = new ConcurrentHashMap<>();
+    private final static ConcurrentHashMap<Class<?>, Map<String, Enum<?>>> ENUMS = new ConcurrentHashMap<>();
+
+    /**
+     * Decode a FaunaDB {@link Value} to a specified type.
+     *
+     * <p>This method is useful if you need to decode
+     * collections types.</p>
+     *
+     * <pre>{@code
+     *     ArrayV array = new ArrayV(Arrays.asList(new StringV("str1"), new StringV("str2")));
+     *
+     *     List<String> listStrings = Decoder.decode(array, Types.arrayListOf(String.class));
+     *     Set<String> setStrings = Decoder.decode(array, Types.hashSetOf(String.class));
+     * }</pre>
+     *
+     * See {@link Types} for more information.
+     *
+     * @param <T> The return type of the method.
+     * @param value The FaunaDB {@link Value} to be decoded.
+     * @param dstType The {@link Type} in which value should be decoded.
+     * @return A instance of type {@link T}
+     */
+    public static <T> T decode(Value value, Type dstType) {
+        return (T) decodeImpl(value, Types.of(dstType));
+    }
+
+    /**
+     * Decode a FaunaDB {@link Value} to a specified type.
+     *
+     * <p>Use this method to decode user defined types like:</p>
+     *
+     * <pre>{@code
+     *     User user = Decoder.decode(new ObjectV(...), User.class);
+     * }</pre>
+     *
+     * <p>It's possible to decode primitive types and array of primitive types:</p>
+     *
+     * <pre>{@code
+     *     String string = Decoder.decode(new StringV("string"), String.class);
+     *     long longValue = Decoder.decode(new LongV(10), long.class);
+     *     int intValue =  Decoder.decode(new LongV(10), int.class);
+     *
+     *     int[] arrayInt = Decoder.decode(new ArrayV(...), int[].class);
+     * }</pre>
+     *
+     * @param <T> The return type of the method.
+     * @param value The FaunaDB {@link Value} to be decoded.
+     * @param dstType The {@link Class} in which value should be decoded.
+     * @return A instance of type {@link T}
+     */
+    public static <T> T decode(Value value, Class<T> dstType) {
+        return (T) decodeImpl(value, Types.of(dstType));
+    }
+
+    static Object decodeImpl(Value value, SimpleType dstType) {
+        Class<?> rawType = dstType.getRawClass();
+
+        if (rawType == Object.class && value == NullV.NULL)
+            return null;
+
+        if (Value.class.isAssignableFrom(rawType))
+            return toValue(value, rawType);
+
+        if (value == null)
+            return null;
+
+        if (rawType.isEnum())
+            return toEnum(value, rawType);
+
+        if (rawType == String.class)
```

Comment:
We have codecs implemented for most of this types, is it possible to implement codecs for the remaining ones and use them as much as the possible? I'm worried with the proliferation of conversion strategies in the same driver.